### PR TITLE
ENHANCEMENT Reduce number of queries used outside of caches

### PR DESCRIPTION
**Pull request by tractorcow**
Parent issue https://github.com/silverstripe/silverstripe-versioned/issues/53

Pretty much every one of the stage / live checks can be built on top of get_versionnumber_by_stage() method, which has its own internal cache.

### CODE REVIEW

File: src/Versioned.php [hunk 1]
```diff
@@ -2203,10 +2191,11 @@ public static function get_versionnumber_by_stage($class, $stage, $id, $cache =
                 self::$cache_versionnumber[$baseClass][$stage] = [];
             }
 
-            self::$cache_versionnumber[$baseClass][$stage][$id] = $version;
+            // Internally store nulls as 0
+            self::$cache_versionnumber[$baseClass][$stage][$id] = $version ?: 0;
```
Comment (flamerohr) [inline src/Versioned.php]: maybe document that `0` means `does not exist` against this variable at the top of the class?

it's a neat approach - but certainly not obvious at first glance.

**Comment by tractorcow** [comment]
Done.

File: src/Versioned.php (hunk 1 above)
Comment (dhensby) [reply src/Versioned.php]: Does a save and/or publish event clear the cache for the record?

Comment (tractorcow) [reply src/Versioned.php]: Yes it does. All writes call flushCache().

**Pull request merged**